File: Algorithm/Python/1000+/3040. Maximum Number of Operations With the Same Score II.py
# Solution 1, backtracking with dfs, TLE
class Solution(object):
    def maxOperations(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        result = [0]
        def dfs(i, j, cnt, _sum):
            result[0] = max(cnt, result[0])
            if i >= j:
                return
            if nums[i]+nums[j] == _sum:
                dfs(i+1, j-1, cnt+1, _sum)
            if nums[i]+nums[i+1] == _sum:
                dfs(i+2, j, cnt+1, _sum)
            if nums[j]+nums[j-1] == _sum:
                dfs(i, j-2, cnt+1, _sum)
        
        n = len(nums)
        dfs(1, n-2, 1, nums[0]+nums[n-1])
        dfs(0, n-3, 1, nums[-2]+nums[-1])
        dfs(2, n-1, 1, nums[0]+nums[1])

        return result[0]

# Solution 2, dfs with cache
from collections import defaultdict

class Solution(object):
    def maxOperations(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n = len(nums)
        memo = defaultdict(int)
        def dfs(i, j, _sum):
            if i >= j:
                return 0
            flag = (_sum, i, j)
            if flag in memo:
                return memo[flag]
            if nums[i]+nums[j] == _sum:
                memo[flag] = max(memo[flag], 1+dfs(i+1, j-1, _sum))
            if nums[i]+nums[i+1] == _sum:
                memo[flag] = max(memo[flag], 1+dfs(i+2, j, _sum))
            if nums[j]+nums[j-1] == _sum:
                memo[flag] = max(memo[flag], 1+dfs(i, j-2, _sum))
            return memo[flag]
        
        return 1+max(dfs(1, n-2, nums[0]+nums[n-1]),
                max(dfs(0, n-3, nums[-2]+nums[-1]),
                    dfs(2, n-1, nums[0]+nums[1]))
        )

File: Algorithm/C++/1000+/3040. Maximum Number of Operations With the Same Score II.cc
#include <vector>
#include <tuple>
#include <unordered_map>

class Solution {
public:
    int maxOperations(std::vector<int>& nums) {
        int n = nums.size();
        std::vector<std::vector<int>> memo1(n, vector<int>(n, -1));
        std::vector<std::vector<int>> memo2(n, vector<int>(n, -1));
        std::vector<std::vector<int>> memo3(n, vector<int>(n, -1));
        return 1 + std::max({
                    dfs(0, n-3, nums[n-2]+nums[n-1], nums, memo1),
                    dfs(1, n-2, nums[0]+nums[n-1], nums, memo2),
                    dfs(2, n-1, nums[0]+nums[1], nums, memo3)
                });
    }

private:
    int dfs(int i, int j, int sum, std::vector<int>& nums, std::vector<std::vector<int>>& memo) {
        if (i >= j) return 0;
        if(memo[i][j] > 0) return memo[i][j];
        int curr=0;
        if (nums[i] + nums[i+1] == sum) {
            curr = 1+dfs(i+2, j, sum, nums, memo);
        }
        if (nums[i] + nums[j] == sum) {
            curr = std::max(curr, 1+dfs(i+1, j-1, sum, nums, memo));
        }
        if (nums[j-1] + nums[j] == sum) {
            curr = std::max(curr, 1+dfs(i, j-2, sum, nums, memo));
        }
        return memo[i][j]=curr;
    }


};